ceph_deploy/gatherkeys.py

```diff
@@ -1,84 +1,262 @@
+            'mon', 'allow *'
+        ]
+    if keytype is 'mds':
+        return [
+            'mon', 'allow profile bootstrap-mds'
+        ]
+    if keytype is 'osd':
+        return [
+            'mon', 'allow profile bootstrap-osd'
+        ]
+    if keytype is 'rgw':
+        return [
+            'mon', 'allow profile bootstrap-rgw'
+        ]
+    raise RuntimeError('Unexpected keytype %s' % keytype)
+
+
+def gatherkeys_missing(args, distro, rlogger, keypath, keytype, dest_dir):
+    """
+    Get or create the keyring from the mon using the mon keyring by keytype and
+    copy to dest_dir
+    """
+    arguments = [
+        '/usr/bin/ceph',
+        '--connect-timeout=25',
+        '--cluster={cluster}'.format(
+            cluster=args.cluster),
+        '--name', 'mon.',
+        '--keyring={keypath}'.format(
+            keypath=keypath),
+        'auth', 'get-or-create',
+        ]
+    identity = keytype_identity(keytype)
+    arguments.append(identity)
+    capabilites = keytype_capabilities(keytype)
+    arguments.extend(capabilites)
+    out, err, code = remoto.process.check(
+        distro.conn,
+        arguments
+        )
+    if code != 0:
+        rlogger.error('"ceph auth get-or-create for keytype %s returned %s' % (keytype, code))
+        for line in err:
+            rlogger.debug(line)
+        return False
+    keyring_name_local = keytype_path_to(args, keytype)
+    keyring_path_local = os.path.join(dest_dir, keyring_name_local)
+    with file(keyring_path_local, 'w') as f:
+        for line in out:
+            f.write(line + '\n')
+    return True
+
+
+def gatherkeys_with_mon(args, host, dest_dir):
+    """
+    Connect to mon and gather keys if mon is in quorum.
+    """
+    distro = hosts.get(host, username=args.username)
+    path_keytype_mon_template = keytype_path_from(args, 'mon')
+    path_keytype_mon = path_keytype_mon_template.format(
+            hostname=host)
+    mon_key = distro.conn.remote_module.get_file(path_keytype_mon)
+    if mon_key is None:
+        LOG.warning("No mon key found. Is '%s' a mon node" % (host))
+        return False
+    mon_name_local = keytype_path_to(args, "mon")
+    mon_path_local = os.path.join(dest_dir, mon_name_local)
+    with file(mon_path_local, 'w') as f:
+        f.write(mon_key)
+    rlogger = logging.getLogger(host)
+    out, err, code = remoto.process.check(
+        distro.conn,
+            [
+                "/usr/bin/ceph",
+                "--connect-timeout=25",
+                "--cluster={cluster}".format(
+                    cluster=args.cluster),
+                "--admin-daemon=/var/run/ceph/{cluster}-mon.{hostname}.asok".format(
+                    hostname=host,
+                    cluster=args.cluster),
+                "mon_status"
+            ]
+        )
+    if code != 0:
+        rlogger.error('"ceph mon_status %s" returned %s' % (host, code))
+        for line in err:
+            rlogger.debug(line)
+        return False
```

Comment:
test_gatherkeys_fail

uses mock of gatherkeys_with_mon on failure.

It checks for RuntimeError.